# Data Cleaning and Preparation

In [107]:
# importing necessary libraries
import pandas as pd
import numpy as np
from IPython.display import display

In [108]:
# Data Cleaning and Preparation
# Load the dataset
df = pd.read_csv('../../data/raw/electric_vehicle_sales_by_makers.csv')

display(df.head())

date vehicle_category         maker  electric_vehicles_sold
0  01-Apr-21       2-Wheelers  OLA ELECTRIC                       0
1  01-Apr-22       2-Wheelers      OKAYA EV                       0
2  01-May-21       2-Wheelers  OLA ELECTRIC                       0
3  01-Jun-21       2-Wheelers  OLA ELECTRIC                       0
4  01-Jul-21       2-Wheelers  OLA ELECTRIC                       0

In [109]:
# Show distinct makers 
print("\n" + 20*'=')
print("Top 10 Distinct makers:")
print(20*'=')
unique_makers = df['maker'].unique()
print('\n'.join([f"- {maker}" for maker in sorted(unique_makers)]))

# Show total makers without limit
print("\n" + 20*'=')
print("Total distinct makers:")
print(20*'=')
print(len(df['maker'].unique()))

# Optional: Show sample data for these makers
print("\n" + 20*'=')
print("Sample data for these makers:")
print(20*'=')
display(df[df['maker'].isin(unique_makers)].head(10))

# Data Shape
print("\n" + 20*'=')
print("Data shape:")
print(20*'=')
print(f"Rows: {df.shape[0]}, Columns: {df.shape[1]}")


Top 10 Distinct makers:
- AMPERE
- ATHER
- BAJAJ
- BATTRE ELECTRIC
- BEING
- BGAUSS
- BMW India
- BYD India
- HERO ELECTRIC
- Hyundai Motor
- JITENDRA
- KIA Motors
- KINETIC GREEN
- MG Motor
- Mahindra & Mahindra
- Mercedes -Benz AG
- OKAYA EV
- OKINAWA
- OLA ELECTRIC
- OTHERS
- PCA Automobiles
- PURE EV
- REVOLT
- TVS
- Tata Motors
- Volvo Auto India

Total distinct makers:
26

Sample data for these makers:


date vehicle_category         maker  electric_vehicles_sold
0  01-Apr-21       2-Wheelers  OLA ELECTRIC                       0
1  01-Apr-22       2-Wheelers      OKAYA EV                       0
2  01-May-21       2-Wheelers  OLA ELECTRIC                       0
3  01-Jun-21       2-Wheelers  OLA ELECTRIC                       0
4  01-Jul-21       2-Wheelers  OLA ELECTRIC                       0
5  01-Aug-21       2-Wheelers  OLA ELECTRIC                       0
6  01-Sep-21       2-Wheelers  OLA ELECTRIC                       0
7  01-Oct-21       2-Wheelers  OLA ELECTRIC                       0
8  01-Nov-21       2-Wheelers  OLA ELECTRIC                       0
9  01-Apr-21       4-Wheelers     BYD India                       0


Data shape:
Rows: 816, Columns: 4


In [110]:
df.groupby(['vehicle_category']).size().reset_index(name='count')


vehicle_category  count
0       2-Wheelers    456
1       4-Wheelers    360

In [111]:
two_wheelers_makers = df[df['vehicle_category'] == '2-Wheelers'] \
  .groupby('maker')['electric_vehicles_sold'] \
  .sum() \
  .sort_values(ascending=False) \
  .reset_index()
  
four_wheelers_makers = df[df['vehicle_category'] == '4-Wheelers'] \
  .groupby('maker')['electric_vehicles_sold'] \
  .sum() \
  .sort_values(ascending=False) \
  .reset_index()


In [112]:
display(df.info())

display(df.describe())

display(df.isnull().sum())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 816 entries, 0 to 815
Data columns (total 4 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   date                    816 non-null    object
 1   vehicle_category        816 non-null    object
 2   maker                   816 non-null    object
 3   electric_vehicles_sold  816 non-null    int64 
dtypes: int64(1), object(3)
memory usage: 25.6+ KB


None

electric_vehicles_sold
count              816.000000
mean              2531.998775
std               4771.077333
min                  0.000000
25%                 42.000000
50%                662.000000
75%               2636.500000
max              44630.000000

date                      0
vehicle_category          0
maker                     0
electric_vehicles_sold    0
dtype: int64

## Conceptualize the Data: Electric Vehicle Makers and Sales Analysis

### Dataset Overview

This dataset captures **electric vehicle sales** by **distinct manufacturers** across India, segmented by vehicle category (`2-Wheelers` and `4-Wheelers`) and monthly sales data. It includes a total of **816 records** with the following key fields:

- `date`: Time period of the sales data (e.g., from April 2021 onward)
- `vehicle_category`: Type of vehicle sold (`2-Wheelers` or `4-Wheelers`)
- `maker`: Manufacturer/brand of the electric vehicle
- `electric_vehicles_sold`: Number of electric vehicles sold by that maker in a given month

### Key Concepts

1. **Manufacturer Market Coverage**: With **26 distinct makers**, the dataset showcases a wide range of players in India’s EV market—from established automotive giants to emerging EV-focused startups.
2. **Vehicle Segment Focus**: A majority of records and sales fall under the `2-Wheelers` category, highlighting the segment's faster adoption curve and consumer preference.
3. **Brand Penetration Over Time**: The dataset supports temporal analysis of how different brands have performed over months and years.
4. **Top Seller Insights**: OLA ELECTRIC, TVS, and ATHER dominate the 2-wheeler market, while Tata Motors leads the 4-wheeler segment.
5. **Early Market Dynamics**: Several makers (e.g., OLA ELECTRIC, OKAYA EV) initially recorded 0 sales—likely due to pre-launch data collection or delayed product rollouts.

### Top Makers by Total Sales (Updated)

#### 🚲 **2-Wheelers:**
The following brands have shown strong performance in electric 2-wheeler sales:

In [113]:
print('Two Wheeler Makers')
display(two_wheelers_makers)

Two Wheeler Makers


maker  electric_vehicles_sold
0      OLA ELECTRIC                  489473
1               TVS                  272575
2             ATHER                  204449
3     HERO ELECTRIC                  170394
4            AMPERE                  167274
5           OKINAWA                  165231
6            OTHERS                  156471
7             BAJAJ                  145606
8            REVOLT                   27828
9          OKAYA EV                   27036
10          PURE EV                   26524
11            BEING                   18392
12           BGAUSS                   15051
13         JITENDRA                   12438
14    KINETIC GREEN                    9585
15  BATTRE ELECTRIC                    4841

Other notable names include BEING, BGAUSS, JITENDRA, and KINETIC GREEN — all contributing to the emerging 2-wheeler EV space.

#### 🚗 **4-Wheelers:**
The 4-wheeler category is more consolidated, with Tata Motors leading by a large margin:

In [114]:
print('Four Wheeler Makers')
display(four_wheelers_makers)

Four Wheeler Makers


maker  electric_vehicles_sold
0          Tata Motors                   88935
1  Mahindra & Mahindra                   41193
2             MG Motor                   13753
3            BYD India                    2419
4        Hyundai Motor                    2076
5      PCA Automobiles                    1684
6            BMW India                    1370
7     Volvo Auto India                     568
8           KIA Motors                     557
9    Mercedes -Benz AG                     388

These figures indicate a **growing interest in EVs among premium and mass-market car manufacturers**, although adoption is still led by a few dominant players.

### Initial Observations

- **Diverse Participation**: With 26 unique manufacturers and both vehicle categories represented, the dataset reflects India's rapidly evolving EV ecosystem.
- **Sales Distribution**: 2-wheelers dominate the EV market by volume, while 4-wheeler sales are led by fewer but established players.
- **Zero Sales Entries**: Many early data points show 0 sales, likely due to vehicles being in pre-launch phases or ramp-up stages.
- **Time-Series Ready**: The dataset spans from April 2021, allowing for monthly trend and growth analysis.
- **Data Quality**: All necessary columns are consistently populated, making the dataset reliable for exploration and visualization.

In [115]:
df.duplicated().sum()

np.int64(0)

## ✅ L – Locate Solvable Issues

### 1. **Check for Missing Values**

**Goal:** Identify if any columns (`date`, `vehicle_category`, `maker`, `electric_vehicles_sold`) have missing/null entries.

```python
df.isnull().sum()
```

✅ If all are zero, great!
❌ If any are non-zero, we’ll handle them in A (Augment).

---

### 2. **Check for Duplicates**

**Goal:** Ensure no exact rows are repeated (same date, maker, category, and sales).

```python
df.duplicated().sum()
```

➡️ If duplicates are found, they can be removed using:

```python
df = df.drop_duplicates()
```

---

### 3. **Inconsistent Maker Names**

**Goal:** Ensure manufacturer names are consistently spelled and formatted. Example issues:

* `"Mercedes -Benz AG"` vs `"Mercedes-Benz AG"`
* `"Mahindra & Mahindra"` vs `"Mahindra and Mahindra"`
* `"Hyundai Motor"` vs `"HYUNDAI"`

```python
df['maker'].unique()
```

➡️ Manually scan and normalize if needed using `.replace()` or `.str.strip().str.upper()`.

---

### 4. **Sales Value Anomalies**

**Goal:** Ensure `electric_vehicles_sold` is always a non-negative integer.

```python
df[df['electric_vehicles_sold'] < 0]
```

➡️ No maker should report negative EV sales. If you find any, those rows need to be cleaned or flagged.

Also check for non-integer or float values:

```python
df[~df['electric_vehicles_sold'].apply(lambda x: isinstance(x, int))]
```

---

### 5. **Date Format Consistency**

**Goal:** Ensure all dates follow the same format (preferably `datetime` type).

```python
df['date'] = pd.to_datetime(df['date'], errors='coerce')
```

Then check for invalid dates:

```python
df[df['date'].isnull()]
```

➡️ If any `NaT` (Not a Time) entries exist, those are fixable or droppable.

---

### 6. **Check for Category Spelling Issues**

**Goal:** Verify `vehicle_category` only has 2 values: `2-Wheelers` and `4-Wheelers`.

```python
df['vehicle_category'].unique()
```

➡️ Typos like `2Wheelers`, `Two Wheelers`, etc., can be standardized:

```python
df['vehicle_category'] = df['vehicle_category'].replace({
    '2Wheelers': '2-Wheelers',
    'Two Wheelers': '2-Wheelers',
    'Four Wheelers': '4-Wheelers'
})
```

---

## 📌 Summary of Solvable Issues You Might Locate:

| Issue                            | How to Detect                          | Solvable With?         |
| -------------------------------- | -------------------------------------- | ---------------------- |
| Missing values                   | `df.isnull().sum()`                    | Fill/drop              |
| Duplicate rows                   | `df.duplicated().sum()`                | `drop_duplicates()`    |
| Inconsistent maker names         | `df['maker'].unique()`                 | `replace()` or mapping |
| Negative or invalid sales values | `df[df['electric_vehicles_sold'] < 0]` | Manual correction/drop |
| Wrong date format or null dates  | `pd.to_datetime()` + `.isnull()`       | `errors='coerce'`      |
| Spelling issues in categories    | `df['vehicle_category'].unique()`      | `replace()`            |

---

Would you like me to generate a Python script that automatically checks and logs these solvable issues for you?
Great! Since you've completed the **C – Conceptualize** part of the **CLEAN** framework, let’s move to **L – Locate Solvable Issues**.

This step is about identifying **realistic data issues** in the dataset that you can **detect and fix** — such as missing values, incorrect formatting, duplicates, inconsistent labels, or unrealistic values.

---

## ✅ L – Locate Solvable Issues

Here’s a structured breakdown for your dataset:

---

### 1. **Check for Missing Values**

**Goal:** Identify if any columns (`date`, `vehicle_category`, `maker`, `electric_vehicles_sold`) have missing/null entries.

```python
df.isnull().sum()
```

✅ If all are zero, great!
❌ If any are non-zero, we’ll handle them in A (Augment).

---

### 2. **Check for Duplicates**

**Goal:** Ensure no exact rows are repeated (same date, maker, category, and sales).

```python
df.duplicated().sum()
```

➡️ If duplicates are found, they can be removed using:

```python
df = df.drop_duplicates()
```

---

### 3. **Inconsistent Maker Names**

**Goal:** Ensure manufacturer names are consistently spelled and formatted. Example issues:

* `"Mercedes -Benz AG"` vs `"Mercedes-Benz AG"`
* `"Mahindra & Mahindra"` vs `"Mahindra and Mahindra"`
* `"Hyundai Motor"` vs `"HYUNDAI"`

```python
df['maker'].unique()
```

➡️ Manually scan and normalize if needed using `.replace()` or `.str.strip().str.upper()`.

---

### 4. **Sales Value Anomalies**

**Goal:** Ensure `electric_vehicles_sold` is always a non-negative integer.

```python
df[df['electric_vehicles_sold'] < 0]
```

➡️ No maker should report negative EV sales. If you find any, those rows need to be cleaned or flagged.

Also check for non-integer or float values:

```python
df[~df['electric_vehicles_sold'].apply(lambda x: isinstance(x, int))]
```

---

### 5. **Date Format Consistency**

**Goal:** Ensure all dates follow the same format (preferably `datetime` type).

```python
df['date'] = pd.to_datetime(df['date'], errors='coerce')
```

Then check for invalid dates:

```python
df[df['date'].isnull()]
```

➡️ If any `NaT` (Not a Time) entries exist, those are fixable or droppable.

---

### 6. **Check for Category Spelling Issues**

**Goal:** Verify `vehicle_category` only has 2 values: `2-Wheelers` and `4-Wheelers`.

```python
df['vehicle_category'].unique()
```

➡️ Typos like `2Wheelers`, `Two Wheelers`, etc., can be standardized:

```python
df['vehicle_category'] = df['vehicle_category'].replace({
    '2Wheelers': '2-Wheelers',
    'Two Wheelers': '2-Wheelers',
    'Four Wheelers': '4-Wheelers'
})
```

---

## 📌 Summary of Solvable Issues You Might Locate:

| Issue                            | How to Detect                          | Solvable With?         |
| -------------------------------- | -------------------------------------- | ---------------------- |
| Missing values                   | `df.isnull().sum()`                    | Fill/drop              |
| Duplicate rows                   | `df.duplicated().sum()`                | `drop_duplicates()`    |
| Inconsistent maker names         | `df['maker'].unique()`                 | `replace()` or mapping |
| Negative or invalid sales values | `df[df['electric_vehicles_sold'] < 0]` | Manual correction/drop |
| Wrong date format or null dates  | `pd.to_datetime()` + `.isnull()`       | `errors='coerce'`      |
| Spelling issues in categories    | `df['vehicle_category'].unique()`      | `replace()`            |

In [116]:
df['vehicle_category'].unique()

array(['2-Wheelers', '4-Wheelers'], dtype=object)

In [117]:
# Replace inconsistent or misspelled maker names
df['maker'] = df['maker'].replace({
    'Mercedes -Benz AG': 'Mercedes-Benz AG',
    'Mahindra & Mahindra': 'Mahindra and Mahindra',
    'HYUNDAI MOTOR': 'HYUNDAI',
    'Bmw India': 'BMW India',
    'kia motors': 'KIA Motors'
})

In [118]:
# Strip extra spaces and make consistent casing (title case)
# df['maker'] = df['maker'].str.strip().str.title()
df['maker'] = df['maker'].str.strip()

In [119]:
df[df['electric_vehicles_sold'] < 0]

Empty DataFrame
Columns: [date, vehicle_category, maker, electric_vehicles_sold]
Index: []

In [120]:
df['date'] = pd.to_datetime(df['date'], errors='coerce')

/tmp/ipykernel_266378/2370506791.py:1: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  df['date'] = pd.to_datetime(df['date'], errors='coerce')


In [121]:
df[df['date'].isnull()]

Empty DataFrame
Columns: [date, vehicle_category, maker, electric_vehicles_sold]
Index: []

In [122]:
df.head()

date vehicle_category         maker  electric_vehicles_sold
0 2021-04-01       2-Wheelers  OLA ELECTRIC                       0
1 2022-04-01       2-Wheelers      OKAYA EV                       0
2 2021-05-01       2-Wheelers  OLA ELECTRIC                       0
3 2021-06-01       2-Wheelers  OLA ELECTRIC                       0
4 2021-07-01       2-Wheelers  OLA ELECTRIC                       0

## ✅ E – Evaluate (and Re-Evaluate) Unsolvable Issues

With the availability of supporting datasets (`ev_sales_by_state.csv` and `ev_sales_enhanced.csv`), some of the earlier limitations can now be addressed:

| Issue                        | Description                                    | Status        | Solution/Next Step                                                         |
| ---------------------------- | ---------------------------------------------- | ------------- | -------------------------------------------------------------------------- |
| ❌ **Missing State Info**     | Original dataset lacks a `state` column        | ✅ Solvable    | Use `ev_sales_by_state.csv` for regional analysis                          |
| ❌ **No Total Sales Info**    | Can't calculate market penetration             | ✅ Solvable    | Use `ev_sales_enhanced.csv` which includes both EV and total vehicle sales |
| ✅ **‘OTHERS’ Maker Group**   | 'OTHERS' category masks individual brands      | ❌ Unsolvable  | Cannot disaggregate unless raw source is available                         |
| ⚠️ **No Vehicle Model Info** | Only brand-level info, no product-level detail | ❌ Unsolvable  | Requires additional columns or new dataset                                 |
| ⚠️ **No Unique ID**          | No primary key per row                         | 🟡 Workaround | Add row index as unique ID if merging or tracking is needed                |

---

### 🧠 Insights:

* **Well-prepared datasets!** I’ve already augmented and extended your data beyond initial assumptions.
* I now have the ability to:
  * Perform **state-wise adoption analysis**
  * Calculate **penetration rates** over time
  * Combine multiple datasets for richer storytelling (great for dashboarding!)

## Adding Unique ID in Each Row

In [123]:
import uuid

df['uuid'] = [uuid.uuid4() for _ in range(len(df))]

display(df.head())

date vehicle_category         maker  electric_vehicles_sold  \
0 2021-04-01       2-Wheelers  OLA ELECTRIC                       0   
1 2022-04-01       2-Wheelers      OKAYA EV                       0   
2 2021-05-01       2-Wheelers  OLA ELECTRIC                       0   
3 2021-06-01       2-Wheelers  OLA ELECTRIC                       0   
4 2021-07-01       2-Wheelers  OLA ELECTRIC                       0   

                                   uuid  
0  465931d3-9bd6-43b5-ab1f-aace65b11dfb  
1  1d7ca2e2-988c-42e9-9d44-8b08a759fbbf  
2  9da18de8-1173-4717-a894-f3b0748e3249  
3  ac373721-db45-46df-bc6a-86f0d663253a  
4  9919575b-4d9d-4823-a9a1-f2a1b7c037c7

## 🔗 A – Augment the Data

**Objective:** Enhance your dataset by **adding new columns, enriching the context**, or **deriving new metrics** from existing fields — which gives you more analytical power and storytelling depth.

---

### 🔎 Dataset: `ev_sales_by_maker`

I already have:

| date | vehicle_category | maker | electric_vehicles_sold |
| ---- | ---------------- | ----- | ---------------------- |

Let’s explore how we can augment this data.

---

## ✅ Ideas to Augment the `ev_sales_by_maker` Dataset

### 1. **Add a Unique ID (Done)**

We’ve already added a unique `id` column — ✅ augmentation started.

---

### 2. **Add Year and Month Columns**

Helpful for time-series grouping and seasonal trends.

```python
df['date'] = pd.to_datetime(df['date'], errors='coerce')
df['year'] = df['date'].dt.year
df['month'] = df['date'].dt.month_name()
```

---

### 3. **Compute Cumulative Sales per Maker**

Tracks how each brand’s sales evolve over time.

```python
df['cumulative_sales'] = df.sort_values(['maker', 'date']) \
                           .groupby('maker')['electric_vehicles_sold'] \
                           .cumsum()
```

---

### 4. **Classify Makers by Segment**

Let’s create a new column — `segment` — to differentiate 2-wheeler and 4-wheeler brands more clearly.

```python
# Basic segmentation by vehicle_category
df['segment'] = df['vehicle_category'].map({
    '2-Wheelers': 'Two-Wheeler Manufacturer',
    '4-Wheelers': 'Four-Wheeler Manufacturer'
})
```

---

### 5. **Add Total Vehicle Share (Optional but powerful)**

If you can merge this with the `ev_sales_enhanced.csv` (which contains total vehicle sales), you can derive:

```python
df['ev_penetration_rate'] = df['electric_vehicles_sold'] / df['total_vehicles_sold']
```

But that requires merging. I can help with that if needed.

---

### Summary: What I Can Add

| New Column Name       | Description                               | Purpose                 |
| --------------------- | ----------------------------------------- | ----------------------- |
| `id`                  | Unique row ID                             | Primary key             |
| `year` & `month`      | Extracted from `date`                     | Time-series grouping    |
| `cumulative_sales`    | Running total per maker                   | Growth tracking         |
| `segment`             | Manufacturer segment (2W vs 4W)           | Categorical insight     |
| `ev_penetration_rate` | EV share among total vehicles (if merged) | Market adoption measure |


In [124]:
df['date'] = pd.to_datetime(df['date'], errors='coerce')
df['year'] = df['date'].dt.year
df['month'] = df['date'].dt.month_name()

In [125]:
df["cumulative_sales"] = (
    df.sort_values(["maker", "date"])
    .groupby("maker")["electric_vehicles_sold"]
    .cumsum()
)

In [126]:
# Basic segmentation by vehicle_category
df['segment'] = df['vehicle_category'].map({
    '2-Wheelers': 'Two-Wheeler Manufacturer',
    '4-Wheelers': 'Four-Wheeler Manufacturer'
})

In [127]:
df_maker = df
df_state = pd.read_csv('../../data/processed/ev_sales_enhanced.csv')
df_state['date'] = pd.to_datetime(df_state['date'], errors='coerce')
df_state.head()

date  year  month  quarter  months_from_start  rolling_mean_ev  \
0 2021-04-01  2021      4        2                  3              0.0   
1 2021-04-01  2021      4        2                  3              9.0   
2 2021-04-01  2021      4        2                  3            436.0   
3 2021-04-01  2021      4        2                  3             26.0   
4 2021-04-01  2021      4        2                  3              0.0   

   ev_growth_rate  ev_penetration  ev_penetration_log  national_market_share  \
0             0.0            0.00            0.000000                   0.00   
1             0.0            5.36            1.850028                   0.14   
2             0.0            0.69            0.524729                   6.90   
3             0.0            0.33            0.285179                   0.41   
4             0.0            0.00            0.000000                   0.00   

   ...  is_at_peak is_missing_ev_growth_rate  \
0  ...           0                         0   
1  ...           1                         0   
2  ...           0                         0   
3  ...           0                         0   
4  ...           0                         0   

   is_missing_segment_preference_ratio is_missing_segment_growth_diff  \
0                                    0                              0   
1                                    0                              0   
2                                    0                              0   
3                                    0                              0   
4                                    0                              0   

   is_missing_regional_avg_penetration  is_missing_adoption_velocity  \
0                                    0                             0   
1                                    0                             0   
2                                    0                             0   
3                                    0                             0   
4                                    0                             0   

                         state  vehicle_category  electric_vehicles_sold  \
0  Andaman and Nicobar Islands        2-Wheelers                       0   
1  Andaman and Nicobar Islands        4-Wheelers                       9   
2               Andhra Pradesh        2-Wheelers                     436   
3               Andhra Pradesh        4-Wheelers                      26   
4            Arunachal Pradesh        2-Wheelers                       0   

   total_vehicles_sold  
0                  515  
1                  168  
2                63444  
3                 7837  
4                 1256  

[5 rows x 41 columns]

In [128]:
import pandas as pd

# Load enhanced data (you can change the file path)
df_state = pd.read_csv('../../data/processed/ev_sales_enhanced.csv')

# Ensure proper date format
df_state['date'] = pd.to_datetime(df_state['date'], errors='coerce')

# Calculate EV Penetration Rate
df_state['ev_penetration_rate'] = df_state['electric_vehicles_sold'] / df_state['total_vehicles_sold']

# Optional: Handle divide-by-zero
df_state['ev_penetration_rate'] = df_state['ev_penetration_rate'].replace([float('inf'), -float('inf')], 0).fillna(0)

# Preview
display(df_state.head())


date  year  month  quarter  months_from_start  rolling_mean_ev  \
0 2021-04-01  2021      4        2                  3              0.0   
1 2021-04-01  2021      4        2                  3              9.0   
2 2021-04-01  2021      4        2                  3            436.0   
3 2021-04-01  2021      4        2                  3             26.0   
4 2021-04-01  2021      4        2                  3              0.0   

   ev_growth_rate  ev_penetration  ev_penetration_log  national_market_share  \
0             0.0            0.00            0.000000                   0.00   
1             0.0            5.36            1.850028                   0.14   
2             0.0            0.69            0.524729                   6.90   
3             0.0            0.33            0.285179                   0.41   
4             0.0            0.00            0.000000                   0.00   

   ...  is_missing_ev_growth_rate is_missing_segment_preference_ratio  \
0  ...                          0                                   0   
1  ...                          0                                   0   
2  ...                          0                                   0   
3  ...                          0                                   0   
4  ...                          0                                   0   

   is_missing_segment_growth_diff is_missing_regional_avg_penetration  \
0                               0                                   0   
1                               0                                   0   
2                               0                                   0   
3                               0                                   0   
4                               0                                   0   

   is_missing_adoption_velocity                        state  \
0                             0  Andaman and Nicobar Islands   
1                             0  Andaman and Nicobar Islands   
2                             0               Andhra Pradesh   
3                             0               Andhra Pradesh   
4                             0            Arunachal Pradesh   

   vehicle_category  electric_vehicles_sold  total_vehicles_sold  \
0        2-Wheelers                       0                  515   
1        4-Wheelers                       9                  168   
2        2-Wheelers                     436                63444   
3        4-Wheelers                      26                 7837   
4        2-Wheelers                       0                 1256   

   ev_penetration_rate  
0             0.000000  
1             0.053571  
2             0.006872  
3             0.003318  
4             0.000000  

[5 rows x 42 columns]

In [130]:
df_maker_month = df_maker.groupby(['date', 'vehicle_category', 'maker'])['electric_vehicles_sold'].sum().reset_index()

# Calculate total EVs sold each month by category
monthly_totals = df_maker_month.groupby(['date', 'vehicle_category'])['electric_vehicles_sold'].sum().reset_index(name='monthly_total_sales')

# Merge and calculate share
df_maker_share = pd.merge(df_maker_month, monthly_totals, on=['date', 'vehicle_category'])
df_maker_share['maker_share_percent'] = df_maker_share['electric_vehicles_sold'] / df_maker_share['monthly_total_sales']

In [133]:
# Load state data
df_state['date'] = pd.to_datetime(df_state['date'], errors='coerce')

# Only keep relevant columns for EV state data
df_state_ev = df_state[['date', 'state', 'vehicle_category', 'electric_vehicles_sold']]

# Rename the column in df_state_ev to avoid confusion
df_state_ev = df_state_ev.rename(columns={'electric_vehicles_sold': 'state_ev_sales'})

# Merge with maker share
df_estimated_maker_state = pd.merge(df_state_ev, df_maker_share, on=['date', 'vehicle_category'], how='inner')

# Estimate how many EVs each maker sold in each state
df_estimated_maker_state['estimated_ev_sales'] = (
    df_estimated_maker_state['state_ev_sales'] * df_estimated_maker_state['maker_share_percent']
).round().astype(int)  # Round to nearest integer since we can't have fractional vehicle sales

In [ ]:
df_estimated_maker_state[df_estimated_maker_state['estimated_ev_sales'] > 1000]

date          state vehicle_category  state_ev_sales  \
2568  2021-07-01      Karnataka       2-Wheelers            3384   
3316  2021-08-01      Karnataka       2-Wheelers            3319   
4064  2021-09-01      Karnataka       2-Wheelers            3539   
4372  2021-09-01     Tamil Nadu       2-Wheelers            2880   
4812  2021-10-01      Karnataka       2-Wheelers            3709   
...          ...            ...              ...             ...   
27608 2024-03-01     Tamil Nadu       2-Wheelers           13350   
27643 2024-03-01  Uttar Pradesh       2-Wheelers            7799   
27644 2024-03-01  Uttar Pradesh       2-Wheelers            7799   
27651 2024-03-01  Uttar Pradesh       2-Wheelers            7799   
27654 2024-03-01  Uttar Pradesh       2-Wheelers            7799   

               maker  electric_vehicles_sold  monthly_total_sales  \
2568   HERO ELECTRIC                    4557                14676   
3316   HERO ELECTRIC                    5527                16050   
4064   HERO ELECTRIC                    6727                18032   
4372   HERO ELECTRIC                    6727                18032   
4812   HERO ELECTRIC                    6799                20813   
...              ...                     ...                  ...   
27608            TVS                   26158               129605   
27643          ATHER                   17012               129605   
27644          BAJAJ                   17716               129605   
27651   OLA ELECTRIC                   44630               129605   
27654            TVS                   26158               129605   

       maker_share_percent  estimated_ev_sales  
2568              0.310507                1051  
3316              0.344361                1143  
4064              0.373059                1320  
4372              0.373059                1074  
4812              0.326671                1212  
...                    ...                 ...  
27608             0.201829                2694  
27643             0.131260                1024  
27644             0.136692                1066  
27651             0.344354                2686  
27654             0.201829                1574  

[425 rows x 9 columns]

## 📊 What This Table Shows

This table is telling a **story about electric vehicle (EV) sales** — specifically:

> "How much each EV company is **estimated** to have sold in **each Indian state**, every month."

We are trying to **guess** how many electric vehicles each company (like OLA, HERO, TVS, etc.) might have sold **in each state** — even though we don’t have that exact number. We estimate it using some clever math.

---

## 🧩 Column-by-Column Explanation

| Column Name              | What It Means (Simple Words)                                                                                              |
| ------------------------ | ------------------------------------------------------------------------------------------------------------------------- |
| `date`                   | 📅 The month and year of sales (e.g., July 2021)                                                                          |
| `state`                  | 🗺️ The Indian state (like Karnataka, Tamil Nadu, etc.)                                                                   |
| `vehicle_category`       | 🚲 Type of EV — 2-Wheelers or 4-Wheelers                                                                                  |
| `state_ev_sales`         | 🚗 How many **EVs** (total of all brands) were sold in that state in that month                                           |
| `maker`                  | 🏢 The EV brand name (like TVS, HERO ELECTRIC, OLA ELECTRIC)                                                              |
| `electric_vehicles_sold` | 📈 How many EVs that brand sold **in total in India** that month                                                          |
| `monthly_total_sales`    | 🧮 How many **EVs were sold by all brands** in India that month (for 2-wheelers)                                          |
| `maker_share_percent`    | 🔢 What % of the national EV sales was from this brand (e.g., HERO had 31% of July 2021 sales)                            |
| `estimated_ev_sales`     | 🔍 A smart guess of **how many EVs that brand sold in that state** that month — we calculate it using the brand’s share % |

---

## 🔍 Let's Break Down One Row

Take this row as an example:

| date       | state     | vehicle\_category | state\_ev\_sales | maker         | electric\_vehicles\_sold | monthly\_total\_sales | maker\_share\_percent | estimated\_ev\_sales |
| ---------- | --------- | ----------------- | ---------------- | ------------- | ------------------------ | --------------------- | --------------------- | -------------------- |
| 2021-07-01 | Karnataka | 2-Wheelers        | 3384             | HERO ELECTRIC | 4557                     | 14676                 | 0.310507 (31%)        | 1051                 |

### 🧠 What this means:

* In **July 2021**, **Karnataka** sold **3384** electric 2-wheelers total (all brands combined).
* In all of India that month, **14,676** electric 2-wheelers were sold.
* HERO ELECTRIC sold **4,557** of those nationally — which is about **31% of India’s total** that month.
* So, we **guess** that HERO ELECTRIC sold **31% of Karnataka’s sales**:

  ```
  estimated_ev_sales = 3384 × 0.310507 ≈ 1051 EVs
  ```

So **1051 EVs in Karnataka** that month were *probably* HERO ELECTRIC vehicles — based on its national market share.

---

## 🔚 Summary

| Question                     | Answer                                                                                                                 |
| ---------------------------- | ---------------------------------------------------------------------------------------------------------------------- |
| 🧐 Why are we estimating?    | Because we don’t have real state-wise brand sales. So we use national brand share to guess.                            |
| 🧠 Is it 100% accurate?      | No, it’s an estimate. But it’s good enough to see trends.                                                              |
| 🔥 What can we do with this? | Make charts like: “Which brand sells the most in each state?”, “Top brands by region”, or “Brand growth by geography”. |


In [139]:
# Create a timestamp for versioning
from datetime import datetime
timestamp = datetime.now().strftime('%Y%m%d')

# 1. Export the main cleaned maker dataset
df_maker.to_csv(f'../../data/processed/ev_sales_by_makers_cleaned_{timestamp}.csv', index=False)

# 2. Export the maker monthly share dataset
df_maker_share.to_csv(f'../../data/processed/ev_maker_monthly_share_{timestamp}.csv', index=False)

# 3. Export the estimated maker state sales dataset
df_estimated_maker_state.to_csv(f'../../data/processed/ev_estimated_maker_state_sales_{timestamp}.csv', index=False)

# 4. Export the enhanced state dataset with penetration rates
df_state.to_csv(f'../../data/processed/ev_sales_by_state_enhanced_{timestamp}.csv', index=False)

# Print confirmation and file info
print("\n=== Export Complete ===")
print(f"Files exported to data/processed/ directory with timestamp: {timestamp}")
print("\nExported Files:")
print(f"1. ev_sales_by_makers_cleaned_{timestamp}.csv")
print(f"   - Rows: {len(df_maker)}")
print(f"   - Columns: {df_maker.columns.tolist()}")

print(f"\n2. ev_maker_monthly_share_{timestamp}.csv")
print(f"   - Rows: {len(df_maker_share)}")
print(f"   - Columns: {df_maker_share.columns.tolist()}")

print(f"\n3. ev_estimated_maker_state_sales_{timestamp}.csv")
print(f"   - Rows: {len(df_estimated_maker_state)}")
print(f"   - Columns: {df_estimated_maker_state.columns.tolist()}")

print(f"\n4. ev_sales_by_state_enhanced_{timestamp}.csv")
print(f"   - Rows: {len(df_state)}")
print(f"   - Columns: {df_state.columns.tolist()}")


=== Export Complete ===
Files exported to data/processed/ directory with timestamp: 20250806

Exported Files:
1. ev_sales_by_makers_cleaned_20250806.csv
   - Rows: 816
   - Columns: ['date', 'vehicle_category', 'maker', 'electric_vehicles_sold', 'uuid', 'year', 'month', 'cumulative_sales', 'segment']

2. ev_maker_monthly_share_20250806.csv
   - Rows: 816
   - Columns: ['date', 'vehicle_category', 'maker', 'electric_vehicles_sold', 'monthly_total_sales', 'maker_share_percent']

3. ev_estimated_maker_state_sales_20250806.csv
   - Rows: 27711
   - Columns: ['date', 'state', 'vehicle_category', 'state_ev_sales', 'maker', 'electric_vehicles_sold', 'monthly_total_sales', 'maker_share_percent', 'estimated_ev_sales']

4. ev_sales_by_state_enhanced_20250806.csv
   - Rows: 2445
   - Columns: ['date', 'year', 'month', 'quarter', 'months_from_start', 'rolling_mean_ev', 'ev_growth_rate', 'ev_penetration', 'ev_penetration_log', 'national_market_share', 'state_rank', 'growth_stage', 'market_concentr